Reorganize interpolator and merger code

eoydvin opened the pull request:
Reorganize base functions as suggested in #15. This solves #38 

Also adds possibility to add individual uncertainty to each CML or rain gauge observation.

eoydvin commented:
One issue is how to treat np.nan observations. The kriging interpolators now does this by ignoring nan inside the lowlevel-functions. The IDW interpolator does not have this functionality, and grid points where a nan value is used for interpolation is set to nan. This is visible the [additiveIDW map in cell marked 18.](https://github.com/OpenSenseAction/mergeplg/blob/4dc1650371ecbb5df660f38116eb9e8a8a94b498/docs/notebooks/openMRG_case_highlevel.ipynb)

My impression is that we would either have to modify the IDW code, so that nans are ignored, or to reinitialize it each time nan observations occurs. Any thoughts @cchwala ?

eoydvin commented:
In the most recent commit I have modified the IDW code to ignor nan, similarly to what is already done in the Kriging functions.

cchwala commented on src/mergeplg/interpolate.py:
```diff
@@ -6,274 +6,444 @@
+
+        return np.concatenate([obs_cmls, obs_gauges]).astype(float)
+
+    def __call__(self, da_grid, da_cmls=None, da_gauges=None):
```
Could we get rid of `da_grid` here. It is never required for the interpolation or checking that the interpolator has to be updated. If the grid would change one needs to create a new interpolator instance.

I see that in `_maybe_update_interpolator()` the grid data is required in some cases (only in `InterpolateOrdinaryKriging`?) to initialize a new interpolator if needed. But we could also store the grid e.g. as `self.xgrid` and `self.ygrid` in `__init__()` and these can then be used.

I also see that there is the case in the merging where we need the gridded data in `__call__()` to find the best match with `nnear` pixels. But there is are new `__call_()` method in the interpolator classes anyway.

IMO removing `da_grid` here would make the code and also using `__call__()` as a user more understandable. But there might be good reasons to keep it that I am just not aware of.

eoydvin replied:
Good. In the updated code the lowlevel interpolator functions get the grid data as numpy arrays xgrid and ygrid. In future updates we could also pass CML and rain gauge coordinates as numpy arrays, instead of xarray.Datasets. That would make the interpolator functions more transferable to potential other projects. I will create an issue.

cchwala replied:
Thanks for the quick reply. I do not fully understand how this is related to removing or keeping `da_grid` as argument in `__call__`. Could you explain that again?

eoydvin replied:
Not directly related, and thinking about it more I think its fine for now. I will not create an issue

cchwala commented on src/mergeplg/interpolate.py:
```diff
@@ -6,274 +6,444 @@
 
-class InterpolateIDW(Base):
-    """Interpolate CML and rain gauge using IDW (CML midpoint)."""
+class Interpolator:
```
I would use the name `InterpolatorBase` here to make it clear that this class is not meant to be used directly.

eoydvin replied:
Fixed

cchwala commented on src/mergeplg/interpolate.py:
```diff
@@ -6,274 +6,444 @@
 
-class InterpolateOrdinaryKriging(Base):
-    """Interpolate CML and radar using neighbourhood ordinary kriging
+class InterpolateKriging(Interpolator):
```
this could also be renamed to `InterpolaeKrigingBase` because it cannot be used directly

cchwala commented on src/mergeplg/merge.py:
```diff
@@ -3,247 +3,462 @@
+                    )
+
+
+class MergeDifferenceIDW(interpolate.InterpolateIDW, MergeBase):
```
Finding the best match within some neariest neighbor pixels is currently not supported here, right?

eoydvin replied:
That is correct. It should be implemented in the future

cchwala reviewed (commented):
Thanks for another large update!

I like how things work now in the notebook. Results also look good and test coverage is almost perfect (and enough now since we have to move forward asap and come back at details of the code later)

I added some smaller comments. Nothing crucial but removing the gridded data from `__call__` in the interpolator would make things cleaner (see my comment there).

I would also have some ideas for restructuring. But these will not change the "user interface" and can be done later, if at all.

eoydvin commented:
> Thanks for another large update!
> 
> I like how things work now in the notebook. Results also look good and test coverage is almost perfect (and enough now since we have to move forward asap and come back at details of the code later)
> 
> I added some smaller comments. Nothing crucial but removing the gridded data from `__call__` in the interpolator would make things cleaner (see my comment there).
> 
> I would also have some ideas for restructuring. But these will not change the "user interface" and can be done later, if at all.

Removed passing the gridded data to the interpolator. Before the interpolated grid copied the coordinates from the radar grid. In the updated version we store the coordinates to self, and the time stamp is extracted from the CML or rain gauge data.

cchwala commented:
Great! 👍

Should I merge now?

eoydvin commented:
Yes, I think it ready to merge
